### DISCUSSION TITLE
Calculation formula

### DISCUSSION THREAD

**Discussion by kajami** [helsenorge/structor#481] [Q&A]
Hello,

I noticed that if you set a Number field in the Structor form builder you can find "Calculation formula" field in Advanced settings.
Can you retrieve input data from previous fields for that calculation formula to use in calculations? For example calculate total points in a form based on the previous questions. Calculation formula would be then something like "question A + question B + question C".

![image](https://github.com/helsenorge/structor/assets/43032284/63c6848a-c3e6-4ac4-8912-d8eaa84388c2)

**Comment by KimHoelHalvorsen** [comment]
In our bulider you can add various fhir-path based formulas. To use them in our end-user UI (reforo https://github.com/helsenorge/refero) you need to set the inputfields as quantity, as well as the sum items. If you use and other renderer, it might suppert other datatypes as well.  

"id": "6124841b-794c-48b9-bd2c-0eba8f7e2ac9",
  "item": [
    {
      "linkId": "251dd3ee-e1d2-46c7-8d5a-ecda88068cbe",
      "type": "group",
      "text": "Calculation",
      "item": [
        {
          "linkId": "832543ea-f43a-49f0-89a3-e7475465780a",
          "type": "display",
          "text": "The data fields need to be Quantity. The sum field need to be Quantity with the following definition: { \"url\": \"http://hl7.org/fhir/StructureDefinition/questionnaire-unit\", \"valueCoding\": { \"code\": \"Score\", \"display\": \"Score\", \"system\": \"http://ehelse.no/Score\" } } And a Fhir-Path: QuestionnaireResponse.descendants().where(linkId='Value1').answer.value.value + QuestionnaireResponse.descendants().where(linkId='Value2').answer.value.value)",
          "required": false,
          "_text": {
            "extension": [
              {
                "url": "http://hl7.org/fhir/StructureDefinition/rendering-markdown",
                "valueMarkdown": "The data fields need to be Quantity.  \nThe sum field need to be Quantity with the following definition:\n\n{ \"url\": \"http://hl7.org/fhir/StructureDefinition/questionnaire-unit\", \"valueCoding\": { \"code\": \"Score\", \"display\": \"Score\", \"system\": \"http://ehelse.no/Score\" } }\n\nAnd a Fhir-Path:\n\nQuestionnaireResponse.descendants().where(linkId='Value1').answer.value.value + QuestionnaireResponse.descendants().where(linkId='Value2').answer.value.value)"
              }
            ]
          }
        },
        {
          "linkId": "Value1",
          "type": "quantity",
          "text": "Value 1",
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
              "valueCoding": {
                "code": "Value",
                "display": "Value",
                "system": "urn:uuid:5b7aa2bd-0af1-4ef0-c5f3-154f64a2ec2f"
              }
            }
          ],
          "required": false
        },
        {
          "linkId": "Value2",
          "type": "quantity",
          "text": "Value 2",
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
              "valueCoding": {
                "code": "Value",
                "display": "Value",
                "system": "urn:uuid:f34e987b-58ef-40af-89f5-c12d3a57eb03"
              }
            }
          ],
          "required": false
        },
        {
          "linkId": "67a20293-da90-42cd-aee5-d934ad1f73d4",
          "type": "quantity",
          "text": "Sum",
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
              "valueCoding": {
                "code": "Score",
                "display": "Score",
                "system": "http://ehelse.no/Score"
              }
            },
            {
              "url": "http://ehelse.no/fhir/StructureDefinition/sdf-calculatedExpression",
              "valueString": "QuestionnaireResponse.descendants().where(linkId='Value1').answer.value.value + QuestionnaireResponse.descendants().where(linkId='Value2').answer.value.value)"
            }
          ],
          "required": false,
          "readOnly": true
        }
      ],
      "required": false
    }
  ]
}


![image](https://github.com/helsenorge/structor/assets/93908741/0b76bc54-6f2c-417e-bbfc-fbedb50ddcf8)
